patternfly/patternfly-elements
# feat: Allow icon sets to be updated

Links to demo pages:
- [Icon demo](https://deploy-preview-867--happy-galileo-ea79c4.netlify.app/elements/pfe-icon/demo/) 
- [Icon panel demo](https://deploy-preview-867--happy-galileo-ea79c4.netlify.app/elements/pfe-icon-panel/demo/) 

---

### Fixes and features

* Update paths to built-in sets: https://github.com/patternfly/patternfly-elements/issues/723
* Storybook icon listing: https://github.com/patternfly/patternfly-elements/issues/728
* Icon panel storybook: https://github.com/patternfly/patternfly-elements/issues/719

### What has changed and why

* Added the ability to override the namespace `rh` or `web` using:
```
PfeIcon.addIconSet(
    "rh",
    "https://random-other-url/svg",
    function resolveIconName(name, iconSetName, iconSetPath) {
      var regex = new RegExp("^" + iconSetName + "-((rh|web)-(.+))");
      var match = regex.exec(name);
      var iconId = match[2] + "-icon-" + match[3].replace("icon-", "");
      return iconSetPath + "/" + iconId + ".svg";
    }
  );
```

#### Smaller updates

* Updated storybook utilities to stop sentence casing dropdown fields
* Add prefixing to all attributes except icon in pfe-icon and pfe-icon-panel
* Create json file to import into stories with all icons listed
* Organize and update format of Sass

**pfe-icon-panel**
  * Prefix slots
  * Set the icon panel to use lg icon when circled and xl when icon is not circled
  * Add schema to icon panel JS
  * Add support for stacked and centered to observed attributes

---

### Testing instructions

Testing this new feature will require running the localhost environment because you need to update in-page JavaScript and view the Network tab changes.

#### Pre-test setup

1. Spin up the local development server in watch mode: `npm run live-demo pfe-icon pfe-icon-panel`.  This will open the demo page and storybook.
2. Navigate to the demo page: http://localhost:8000/elements/pfe-icon/demo/
3. Inspect the page and select the Network tab.  Refresh the page.
4. Filter results by `Img` and check the URL source of the SVGs being loaded.  They should point to: https://access.redhat.com/webassets/avalon/j/lib/rh-iconfont-svgs/
5. Open your code editor and navigate to `./elements/pfe-icon/demo/index.html`
6. Locate the `PfeIcon.addIconSet` function on line 38:
  ```
  PfeIcon.addIconSet(
        "local",
        "./",
        function(name, iconSetName, iconSetPath) {
          var regex = new RegExp(`^${iconSetName}-(.*)`);
          var match = regex.exec(name);
          return iconSetPath + match[1] + ".svg";
        }
      );
  ```

### Test 1: Update all inputs to the addIconSet function
Update the function to:
  ```
  PfeIcon.addIconSet(
        "rh",
        "./",
        function(name, iconSetName, iconSetPath) {
          return "square.svg";
      });
  ```
  Upon refresh, you should see: 
<img width="1243" alt="Screen Shot 2020-05-21 at 2 30 38 PM" src="https://user-images.githubusercontent.com/1840295/82592962-b0e9aa00-9b6f-11ea-8204-406fa2da0b83.png">
Check the Network tab and you should see only 1 SVG fetched against localhost for square.svg for the rh icons.  Web icons will continue to fetch from access and render correctly as before.


### Test 2: Update only the path for the rh namespace
Update the function to:
  ```
  PfeIcon.addIconSet("rh", "./");
  ```
  Upon refresh, you should see: 
<img width="1247" alt="Screen Shot 2020-05-21 at 2 31 44 PM" src="https://user-images.githubusercontent.com/1840295/82593072-e2fb0c00-9b6f-11ea-8b7c-9b77fac912b0.png">
  Check the Network tab and you should see a lot of 404 responses querying for icons against localhost.  Web icons will continue to fetch from access and render correctly as before.

#### Browser requirements

Your component should work in all of the following environments:


### Ready-for-merge Checklist

Check off items as they are completed.  Feel free to delete items if they are not applicable to your PR.

- [x] [Demo / walkthrough](https://drive.google.com/file/d/1I8Esz7y6Pr8aDccynfoPwQK9SYnK4P_H/view?usp=sharing)


**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**
